ui/src/BaseEditor.ts

```diff
@@ -0,0 +1,146 @@
+        const filePath: string | undefined = Object.keys(this.fileMap).find((fsPath: string) => path.relative(doc.fsPath, fsPath) === '');
+        if (filePath) {
+            const [textDocument, context]: [vscode.TextDocument, ContextT] = this.fileMap[filePath];
+            await this.updateRemote(context, textDocument);
+        }
+    }
+
+    public async dispose(): Promise<void> {
+        // tslint:disable-next-line:no-unsafe-any
+        Object.keys(this.fileMap).forEach(async (key: string) => await fse.remove(path.dirname(key)));
+    }
+
+    public async onDidSaveTextDocument(globalState: vscode.Memento, doc: vscode.TextDocument): Promise<void> {
+        const filePath: string | undefined = Object.keys(this.fileMap).find((fsPath: string) => path.relative(doc.uri.fsPath, fsPath) === '');
+        if (!this.ignoreSave && filePath) {
+            const context: ContextT = this.fileMap[filePath][1];
+            const showSaveWarning: boolean | undefined = vscode.workspace.getConfiguration().get(this.showSavePromptKey);
+            let shouldUpdateRemote: boolean = false;
+
+            if (showSaveWarning) {
+                const message: string = await this.getSaveConfirmationText(context);
+                const result: vscode.MessageItem | undefined = await vscode.window.showWarningMessage(message, DialogResponses.upload, DialogResponses.dontWarn, DialogResponses.dontUpload);
+
+                if (result === DialogResponses.upload) {
+                    shouldUpdateRemote = true;
+                } else if (result === DialogResponses.dontWarn) {
+                    await vscode.workspace.getConfiguration().update(this.showSavePromptKey, false, vscode.ConfigurationTarget.Global);
+                    await globalState.update(this.showSavePromptKey, true);
+                    shouldUpdateRemote = true; // upload and don't warm me again
+                } else {
+                    throw new UserCancelledError();
+                }
+            } else {
+                shouldUpdateRemote = true;
+            }
+
+            if (shouldUpdateRemote) {
+                await this.updateRemote(context, doc);
+            }
+        }
+    }
+
+    protected appendToOutput(value: string): void {
+        if (!!this.outputChannel) {
+            this.outputChannel.append(value);
+            this.outputChannel.show(true);
+        }
+    }
+
+    protected appendLineToOutput(value: string): void {
+        if (!!this.outputChannel) {
+            this.outputChannel.appendLine(value);
+            this.outputChannel.show(true);
+        }
+    }
+
+    private async updateRemote(context: ContextT, doc: vscode.TextDocument): Promise<void> {
+        const filename: string = await this.getFilename(context);
+        this.appendToOutput(localize('updating', 'Updating "{0}" ...', filename));
+        const updatedData: string = await this.updateData(context, doc.getText());
+        this.appendLineToOutput(localize('done', ' Done.'));

```

Comment:
Here's another case - don't write 'done' without saying which 'file' is done. Otherwise it gets confusing if I'm updating multiple files at a time